File: packages/MeeshySDK/Sources/MeeshySDK/Models/MediaCrop.swift
import CoreGraphics
import Foundation

/// **Un recadrage est une BORNE, jamais un ré-encodage** (#5085, vue `2d`/`4c`).
///
/// > « ces trois gestes écrivent des bornes · aucun ne ré-encode · aucun
/// > n'invalide la montée » — planche `4c`
///
/// C'est cette phrase qui rend le recadrage décidable, et elle dicte la FORME
/// du type : un rectangle NORMALISÉ posé sur la source, en fractions de 0 à 1.
/// Des pixels auraient lié la borne à une résolution — donc au fichier — et
/// c'est exactement ce que la planche interdit : le fichier doit pouvoir partir
/// pendant que l'auteur compose, et continuer de partir quand il recadre.
///
/// La jumelle temporelle existe déjà et fonctionne : `sourceStart`/`sourceEnd`
/// bornent la LECTURE sans toucher au fichier. Ce type borne le CADRE de la
/// même façon.
public struct MediaCropRect: Codable, Equatable, Hashable, Sendable {
    /// Fractions de la source, origine en haut à gauche.
    public var x: Double
    public var y: Double
    public var width: Double
    public var height: Double

    public init(x: Double, y: Double, width: Double, height: Double) {
        self.x = x
        self.y = y
        self.width = width
        self.height = height
    }

    /// Le cadre ENTIER — l'absence de recadrage, écrite.
    public static let full = MediaCropRect(x: 0, y: 0, width: 1, height: 1)

    public var isFull: Bool { self == .full }
}

/// Les proportions que la vue `2d` offre — `9:16 · 4:5 · 1:1 · LIBRE`.
public enum MediaCropRatio: String, CaseIterable, Sendable {
    case portrait916
    case portrait45
    case square
    case free

    /// Le rapport largeur/hauteur VISÉ. `nil` pour `LIBRE`, qui n'en impose
    /// aucun — et c'est la seule des quatre qui laisse l'auteur décider.
    public var value: Double? {
        switch self {
        case .portrait916: return 9.0 / 16.0
        case .portrait45:  return 4.0 / 5.0
        case .square:      return 1
        case .free:        return nil
        }
    }
}

public enum MediaCropRule {

    /// **Ramène un rectangle DANS la source.** Une borne qui déborde ne casse
    /// rien au rendu — le moteur clippe — mais elle voyage jusqu'aux deux
    /// autres clients et jusqu'au renderer d'export, où un décodeur plus strict
    /// ou une multiplication par la taille réelle peut en faire autre chose.
    /// Même raison que la normalisation de `StoryMediaRotation`.
    ///
    /// **L'ORIGINE est bornée pour que le plancher TIENNE.** L'écriture
    /// naïve — origine bornée à `1`, puis dimension bornée à `1 - origine` —
    /// laisse la seconde borne DÉFAIRE la première : à `y == 1`,
    /// `min(max(0.01, h), 0)` rend `0`, c'est-à-dire exactement le média
    /// invisible que le plancher existe pour empêcher.
    ///
    /// Le défaut a vécu ici jusqu'au 2026-09-04, et c'est le portage
    /// TypeScript (#5085) qui l'a trouvé : le témoin qui l'attrape doit porter
    /// sur une origine EN DÉBORDEMENT, et les témoins Swift n'éprouvaient que
    /// des rectangles valides — où les deux écritures s'accordent.
    public static func clamped(_ rect: MediaCropRect) -> MediaCropRect {
        let place = 1 - minimumSide
        let l = min(max(0, rect.x), place)
        let h = min(max(0, rect.y), place)
        // Une largeur nulle rendrait un média INVISIBLE sans rien signaler ;
        // le plancher garde une bande étroite plutôt qu'un vide.
        let w = min(max(minimumSide, rect.width), 1 - l)
        let t = min(max(minimumSide, rect.height), 1 - h)
        return MediaCropRect(x: l, y: h, width: w, height: t)
    }

    /// Un pour cent de la source. En dessous, l'auteur ne voit plus ce qu'il
    /// cadre, et le geste devient un piège plutôt qu'un outil.
    public static let minimumSide: Double = 0.01

    /// **Le plus grand rectangle du rapport VISÉ qui tient dans la source**,
    /// centré. C'est ce qu'une pastille de proportion doit poser : l'auteur
    /// choisit une forme, pas un cadrage — il déplacera ensuite s'il veut.
    ///
    /// - Parameter sourceRatio: largeur/hauteur de la SOURCE. Sans lui, un
    ///   9:16 posé sur une photo panoramique et sur un portrait rendrait le
    ///   même rectangle en fractions — donc deux formes différentes à l'écran.
    public static func centered(ratio: MediaCropRatio,
                                sourceRatio: Double) -> MediaCropRect {
        guard let cible = ratio.value, sourceRatio > 0 else { return .full }
        // En fractions de la source, un rectangle de rapport `cible` a
        // w/h × sourceRatio == cible. On garde la plus grande des deux
        // dimensions à 1 et on déduit l'autre.
        let w: Double
        let h: Double
        if cible >= sourceRatio {
            w = 1
            h = min(1, sourceRatio / cible)
        } else {
            h = 1
            w = min(1, cible / sourceRatio)
        }
        return clamped(MediaCropRect(x: (1 - w) / 2, y: (1 - h) / 2, width: w, height: h))
    }

    /// **Ce que le recadrage rend au rendu** — le rapport EFFECTIF de l'objet.
    /// Un média recadré n'a plus les proportions de son fichier, et c'est ce
    /// nombre que la carte doit ajuster, pas `aspectRatio`.
    public static func effectiveRatio(sourceRatio: Double, crop: MediaCropRect?) -> Double {
        guard let crop, !crop.isFull, crop.height > 0 else { return sourceRatio }
        return sourceRatio * (crop.width / crop.height)
    }
}

File: packages/shared/utils/media-crop.ts
/**
 * **Le recadrage d'un média de scène — la loi, une fois, pour les trois
 * clients** (#5085).
 *
 * ## Ce qui a rendu ce fichier nécessaire
 *
 * Le modèle est né côté iOS (`MediaCropRect` / `MediaCropRule`, Swift) et
 * traversait la passerelle SANS AUCUN LECTEUR : `crop` voyage dans
 * `payload`, que `canvas-v3.ts` déclare `z.record(z.string(), z.unknown())`
 * — permissif PAR CONTRAT. La clé passait donc la validation, arrivait chez
 * le web et chez Android, et n'y était lue par personne.
 *
 * **Une image recadrée sur iOS se rendait ENTIÈRE ailleurs**, sans qu'un seul
 * test ne rougisse : un schéma permissif n'a pas de site où refuser, et un
 * lecteur qui ignore un champ ne se distingue pas d'un lecteur qui ne l'a
 * jamais reçu.
 *
 * > C'est la forme du § « une énumération de sites porte DEUX affirmations »
 * > du `CLAUDE.md` : le lot iOS savait dire « ces sites appliquent la règle »,
 * > il ne savait pas dire « ce sont les sites où la règle s'applique ».
 *
 * ## Le contrat de fil
 *
 * Quatre nombres, en FRACTIONS de la source, écrits par
 * `CanvasV3Migration` (Swift) sous `cropX` · `cropY` · `cropW` · `cropH`, et
 * OMIS quand le recadrage est plein. L'absence est donc un fait lisible : pas
 * de recadrage, pas de clé — jamais `{0,0,1,1}` écrit explicitement.
 *
 * Les quatre se lisent ENSEMBLE ou pas du tout. Un recadrage amputé d'une
 * borne n'a pas de repli sensé : compléter par un défaut fabriquerait un
 * cadrage que personne n'a posé, et le rendrait indiscernable d'un vrai.
 */

export type MediaCropRect = {
  readonly x: number;
  readonly y: number;
  readonly width: number;
  readonly height: number;
};

/** Le recadrage neutre — toute la source. */
export const FULL_MEDIA_CROP: MediaCropRect = { x: 0, y: 0, width: 1, height: 1 };

/**
 * Un pour cent de la source. En dessous, l'auteur ne voit plus ce qu'il cadre.
 * Le nombre est le même que `MediaCropRule.minimumSide` côté Swift — deux
 * planchers différents feraient deux bandes différentes pour un même geste.
 */
export const MINIMUM_CROP_SIDE = 0.01;

export const isFullMediaCrop = (crop: MediaCropRect): boolean =>
  crop.x === 0 && crop.y === 0 && crop.width === 1 && crop.height === 1;

/**
 * **Ramène un rectangle DANS la source.** Une borne qui déborde ne casse rien
 * au rendu — le moteur clippe — mais elle voyage jusqu'au renderer d'export,
 * où une multiplication par la taille réelle peut en faire autre chose.
 *
 * **L'ORIGINE est bornée pour que le plancher TIENNE.** Écrite naïvement —
 * origine bornée à `1`, puis dimension bornée à `1 - origine` — la seconde
 * borne DÉFAIT la première : à `y = 1`, `min(max(0,01, h), 0)` rend `0`,
 * c'est-à-dire exactement le média invisible que le plancher existe pour
 * empêcher. Le témoin qui l'attrape doit donc porter sur une origine EN
 * DÉBORDEMENT ; sur un rectangle valide, les deux écritures s'accordent.
 *
 * Miroir de `MediaCropRule.clamped` (Swift), qui portait le même défaut et a
 * été corrigé dans le même lot — deux planchers qui ne tiennent pas au même
 * endroit rendraient deux bandes différentes pour un même geste.
 */
export const clampMediaCrop = (crop: MediaCropRect): MediaCropRect => {
  const room = 1 - MINIMUM_CROP_SIDE;
  const x = Math.min(Math.max(0, crop.x), room);
  const y = Math.min(Math.max(0, crop.y), room);
  return {
    x,
    y,
    width: Math.min(Math.max(MINIMUM_CROP_SIDE, crop.width), 1 - x),
    height: Math.min(Math.max(MINIMUM_CROP_SIDE, crop.height), 1 - y),
  };
};

const finite = (value: unknown): number | null =>
  typeof value === 'number' && Number.isFinite(value) ? value : null;

/**
 * **Lit le recadrage d'un payload d'objet canvas-v3.**
 *
 * Rend `null` — « pas de recadrage » — dans TROIS cas qui sont le même fait :
 * aucune clé, des clés partielles, ou un recadrage plein. Les distinguer chez
 * l'appelant l'obligerait à connaître la forme de fil, ce que ce module existe
 * pour lui épargner.
 */
export const readMediaCrop = (
  payload: Record<string, unknown> | null | undefined,
): MediaCropRect | null => {
  if (!payload) return null;
  const x = finite(payload.cropX);
  const y = finite(payload.cropY);
  const width = finite(payload.cropW);
  const height = finite(payload.cropH);
  if (x === null || y === null || width === null || height === null) return null;
  const crop = clampMediaCrop({ x, y, width, height });
  return isFullMediaCrop(crop) ? null : crop;
};

/**
 * **Ce que le recadrage rend au rendu** — le rapport EFFECTIF de l'objet.
 * Un média recadré n'a plus les proportions de son fichier, et c'est ce nombre
 * que la carte doit ajuster, jamais `aspectRatio`.
 */
export const effectiveMediaRatio = (
  sourceRatio: number,
  crop: MediaCropRect | null | undefined,
): number => {
  if (!crop || isFullMediaCrop(crop) || crop.height <= 0) return sourceRatio;
  return sourceRatio * (crop.width / crop.height);
};

/**
 * **Le recadrage en style CSS, pour un média qui remplit son conteneur.**
 *
 * Le web n'a pas d'équivalent de `CALayer.contentsRect` : la seule façon de
 * montrer une FRACTION d'une image sans la ré-encoder est de l'agrandir puis
 * de la décaler sous un conteneur qui coupe. C'est exactement ce que fait
 * `contentsRect` en interne, et c'est pourquoi le résultat est identique — le
 * pixel n'est jamais retouché, ni ici ni là.
 *
 * Le décalage se dit en POURCENTAGE de background-position, dont la
 * convention n'est pas linéaire : `p%` aligne le point `p%` de l'image sur le
 * point `p%` du conteneur. Le point visé est donc `x / (1 - width)`, et non
 * `x` — la division est le piège de cette conversion, et elle est indéfinie
 * quand la bande occupe toute la dimension (`width === 1`), auquel cas il n'y
 * a rien à décaler.
 *
 * Rendu pour un conteneur en `overflow: hidden` ; le média porte
 * `position: absolute` et ce style.
 */
export type MediaCropStyle = {
  readonly width: string;
  readonly height: string;
  readonly left: string;
  readonly top: string;
};

export const mediaCropStyle = (crop: MediaCropRect): MediaCropStyle => {
  const bounded = clampMediaCrop(crop);
  const pct = (value: number) => `${value * 100}%`;
  return {
    width: pct(1 / bounded.width),
    height: pct(1 / bounded.height),
    left: pct(-bounded.x / bounded.width),
    top: pct(-bounded.y / bounded.height),
  };
};
